Remove .pyi from /_multi_tensor

izdeby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #47848 [WIP] Clean up backward compatibility list
* #46664 Move device guard from MultiTensorApply.cuh
* **#46663 Remove .pyi from /_multi_tensor**

In this PR:
Following up on previous [feedback](https://github.com/pytorch/pytorch/pull/45408#issuecomment-702913146).
Fixing all the mypy issues that are coming up with removal of .pyi files

Differential Revision: [D24453344](https://our.internmc.facebook.com/intern/diff/D24453344)

Commit:
Remove .pyi from /_multi_tensor

[ghstack-poisoned]

Commit:
Update on "Remove .pyi from /_multi_tensor"


Following up on previous [feedback](https://github.com/pytorch/pytorch/pull/45408#issuecomment-702913146)

[ghstack-poisoned]

rgommers commented:
@izdeby the `test_optim` failures in `pytorch-linux-bionic-rocm3.8-py3.6-test2` look related, there are several `_multi_tensor` failures.

rgommers commented on torch/optim/optimizer.py:
```diff
@@ -1,11 +1,15 @@
 import warnings
 
+import torch
+from .. import Tensor
```
minor: you may just use `Tensor = torch.Tensor` or use `torch.Tensor` on line 12 instead, a little faster.

rgommers reviewed (commented):
Overall LGTM, thanks @izdeby. Relevant failures need fixing.

Commit:
Update on "Remove .pyi from /_multi_tensor"


Following up on previous [feedback](https://github.com/pytorch/pytorch/pull/45408#issuecomment-702913146)

Differential Revision: [D24453344](https://our.internmc.facebook.com/intern/diff/D24453344)

[ghstack-poisoned]

Commit:
Update on "Remove .pyi from /_multi_tensor"


Following up on previous [feedback](https://github.com/pytorch/pytorch/pull/45408#issuecomment-702913146)

Differential Revision: [D24453344](https://our.internmc.facebook.com/intern/diff/D24453344)

[ghstack-poisoned]

izdeby committed:
Update on "Remove .pyi from /_multi_tensor"


Following up on previous [feedback](https://github.com/pytorch/pytorch/pull/45408#issuecomment-702913146)

Differential Revision: [D24453344](https://our.internmc.facebook.com/intern/diff/D24453344)

[ghstack-poisoned]

Commit:
Update on "Remove .pyi from /_multi_tensor"


Following up on previous [feedback](https://github.com/pytorch/pytorch/pull/45408#issuecomment-702913146)

Differential Revision: [D24453344](https://our.internmc.facebook.com/intern/diff/D24453344)

[ghstack-poisoned]

Commit:
Update on "Remove .pyi from /_multi_tensor"


Following up on previous [feedback](https://github.com/pytorch/pytorch/pull/45408#issuecomment-702913146)

Differential Revision: [D24453344](https://our.internmc.facebook.com/intern/diff/D24453344)

[ghstack-poisoned]

Commit:
Update on "Remove .pyi from /_multi_tensor"


Following up on previous [feedback](https://github.com/pytorch/pytorch/pull/45408#issuecomment-702913146)

Differential Revision: [D24453344](https://our.internmc.facebook.com/intern/diff/D24453344)

[ghstack-poisoned]

zou3519 commented on mypy.ini:
```diff
@@ -218,6 +218,33 @@ ignore_errors = True
 [mypy-torch.overrides]
 ignore_errors = True
 
+[mypy-torch.optim._multi_tensor]
+ignore_errors = True
+
+[mypy-torch.optim._multi_tensor.adam]
+ignore_errors = True
+
+[mypy-torch.optim._multi_tensor.adamw]
+ignore_errors = True
+
+[mypy-torch.optim._multi_tensor.sgd]
+ignore_errors = True
+
+[mypy-torch.optim._multi_tensor.rmsprop]
+ignore_errors = True
+
+[mypy-torch.optim._multi_tensor.rprop]
+ignore_errors = True
+
+[mypy-torch.optim._multi_tensor.asgd]
+ignore_errors = True
+
+[mypy-torch.optim._multi_tensor.adamax]
+ignore_errors = True
+
+[mypy-torch.optim._multi_tensor.adadelta]
+ignore_errors = True
```
The typing changes look fine, but why do we ignore errors here?

gchanan commented:
why?

Commit:
Update on "Remove .pyi from /_multi_tensor"


In this PR:
Following up on previous [feedback](https://github.com/pytorch/pytorch/pull/45408#issuecomment-702913146)

Differential Revision: [D24453344](https://our.internmc.facebook.com/intern/diff/D24453344)

[ghstack-poisoned]

Commit:
Update on "Remove .pyi from /_multi_tensor"


In this PR:
Following up on previous [feedback](https://github.com/pytorch/pytorch/pull/45408#issuecomment-702913146)

Differential Revision: [D24453344](https://our.internmc.facebook.com/intern/diff/D24453344)

[ghstack-poisoned]

Commit:
Update on "Remove .pyi from /_multi_tensor"


In this PR:
Following up on previous [feedback](https://github.com/pytorch/pytorch/pull/45408#issuecomment-702913146).
Fixing all the mypy issues that are coming up with removal of .pyi files

Differential Revision: [D24453344](https://our.internmc.facebook.com/intern/diff/D24453344)

[ghstack-poisoned]

Commit:
Update on "Remove .pyi from /_multi_tensor"


In this PR:
Following up on previous [feedback](https://github.com/pytorch/pytorch/pull/45408#issuecomment-702913146).
Fixing all the mypy issues that are coming up with removal of .pyi files

Differential Revision: [D24453344](https://our.internmc.facebook.com/intern/diff/D24453344)

[ghstack-poisoned]

izdeby commented on mypy.ini:
```diff
@@ -356,3 +356,30 @@ ignore_missing_imports = True
 
 [mypy-librosa.*]
 ignore_missing_imports = True
+
+[mypy-torch.optim._multi_tensor]
+ignore_errors = True
+
+[mypy-torch.optim._multi_tensor.adam]
+ignore_errors = True
+
+[mypy-torch.optim._multi_tensor.adamw]
+ignore_errors = True
+
+[mypy-torch.optim._multi_tensor.sgd]
+ignore_errors = True
+
+[mypy-torch.optim._multi_tensor.rmsprop]
+ignore_errors = True
+
+[mypy-torch.optim._multi_tensor.rprop]
+ignore_errors = True
+
```
@zou3519, sadly thats not a simple fix. Updating [this line](https://github.com/pytorch/pytorch/blob/275a89a7ee45772b5d353ec488ec2cd0c78a69c3/tools/pyi/gen_pyi.py#L135) will trigger a lot of other errors. There are a lot of special casing around mypy and especially for the methods that return list of tensors. 

I will follow up on this if thats ok.

zou3519 replied:
Yup, follow up is fine

zou3519 commented on torch/optim/_multi_tensor/sgd.py:
```diff
@@ -52,8 +53,9 @@ class SGD(Optimizer):
 
-    def __init__(self, params, lr=required, momentum=0, dampening=0,
-                 weight_decay=0, nesterov=False):
+    def __init__(self, params: _params_t, lr: float = required,
```
This passes the tests, but we probably don't want to change this to using a float (I don't know the downstream consequences of this and this looks like a pre-existing problem with the sgd type hints) Would prefer to revert this chance since we're not handling the mypy issues anyways

zou3519 commented on torch/optim/_multi_tensor/adamax.py:
```diff
@@ -96,7 +98,7 @@ def step(self, closure=None):
                     grad.abs().add_(eps).unsqueeze_(0)
                 ], 0)
-                torch.max(norm_buf, 0, keepdim=False, out=(exp_inf, exp_inf.new().long()))
+                torch.amax(norm_buf, 0, keepdim=False, out=exp_inf)
```
Are there any behavior changes here?

ngimel replied:
No behavior changes, this change is picked from regular optimizer. Before we had `amax`, we were forced to use `max` function that returned indices, and created a dummy tensor to hold indices, that's not necessary now.

zou3519 reviewed (commented):
Two things that looked concerning to me, but otherwise everything else looks good

izdeby commented:
Closing for now due to long list of TODO items that this PR will create. Should come back and do the whole thing instead
